CS483/CS483/Kartaclysm/StateMachine/Gameplay/StatePaused.cpp

```diff
@@ -0,0 +1,166 @@
+
+	if (m_pPauseDelegate != nullptr)
+	{
+		// If this is the first frame suspended: remove listener
+		HeatStroke::EventManager::Instance()->RemoveListener("Pause", m_pPauseDelegate);
+		delete m_pPauseDelegate;
+		m_pPauseDelegate = nullptr;
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:		Unsuspend
+// Parameter:	const int p_iPrevState - index of previous state popped
+// 
+// Called when this state is popped back to top of stack.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::Unsuspend(const int p_iPrevState)
+{
+	m_bSuspended = false;
+	
+	if (m_pPauseDelegate == nullptr)
+	{
+		m_pPauseDelegate = new std::function<void(const HeatStroke::Event*)>(std::bind(&StatePaused::PauseGame, this, std::placeholders::_1));
+		HeatStroke::EventManager::Instance()->AddListener("Pause", m_pPauseDelegate);
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:    Update
+// Parameter: const float p_fDelta
+// 
+// Called each from when this state is active.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::Update(const float p_fDelta)
+{
+	// Do not update when suspended
+	if (!m_bSuspended)
+	{
+		m_pGameObjectManager->Update(p_fDelta);
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:    PreRender
+// 
+// Called before rendering occurs.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::PreRender()
+{
+	// Render even when suspended
+	m_pGameObjectManager->PreRender();
+}
+
+//------------------------------------------------------------------------------
+// Method:    Exit
+// 
+// Called when this state is popped off the stack.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::Exit()
+{
+	m_bSuspended = false;
+
+	if (m_pPauseDelegate != nullptr)
+	{
+		HeatStroke::EventManager::Instance()->RemoveListener("Pause", m_pPauseDelegate);
+		delete m_pPauseDelegate;
+		m_pPauseDelegate = nullptr;
+	}
+
+	if (m_pGameObjectManager != nullptr)
+	{
+		m_pGameObjectManager->DestroyAllGameObjects();
+		delete m_pGameObjectManager;
+		m_pGameObjectManager = nullptr;
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:    PauseGame
+// Parameter: const HeatStroke::Event* p_pEvent - Event that triggers when a player pauses
+// 
+// Unpause the game by popping the Pause State.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::PauseGame(const HeatStroke::Event* p_pEvent)
+{
+	// Get the player who paused the game
+	int iPlayer = 0;
+	p_pEvent->GetOptionalIntParameter("Player", iPlayer, iPlayer);
+
+	// Pop current state
+	printf("Unpausing\n");
```

Comment:
What's currently stopping you from having a boolean along the lines of m_bPauseInputPressedLastFrame (though a better name would be welcome) that is initially set to true when the state is entered, and set to false when the previous value was true and the pause button is not currently pressed inside StatePaused::Update().  Then check that value here to see if you need to exit.

It seems like a trivial fix, so why are we putting up with the issue of thrashing between states when we know the fix?  If I'm missing something, let me know.